### PR TITLE
Migrate dbus-glib to gdbus.

### PR DESCRIPTION
**Pull request by yetist**
Please test and review it, thank you.

### CODE REVIEW

**Comment by raveit65** [comment]
Indents can be improved in a lot of places or not?
Looks like we have a problem with using different 'tab sizes` in editors.
Personal, i don't add new tabs in code and i replace tabs with spaces in every part where i do code changes, to avoid the tab size problem.
And why last commit?
In firefox browser old  indents are looking better than new ones.
https://github.com/mate-desktop/mate-panel/pull/864/commits/98dff560a8f2f503834dfaccd29297619b05deb6
In pluma it looks complete different.
pluma-example
Please do not replace a mixed tab+spaces line with tabs.
Spaces is always the better choice.

Ok, i start testing now.

**Comment by rcaridade145** [comment]
@raveit65 what i've noticed in pluma is that sometimes i have to press tab twice for it to visually display a tab. However if  i save and open with nano or vi i'll have 2 tabs.

**Comment by raveit65** [comment]
I see von different how often i have to press the tab key with using pluma or vi.
Problem is that the tab size can be different in editors.
4 spaces =1 tab,  8 spaces =1 tab, 16 spaces =1 tab, 1024 spaces =1 tab........
For this reason i try to avoid using tabs.

**Comment by lukefromdc** [comment]
Any progress on this? it is being held up over the tab/space issue. Should I fix that myself and open a new PR?

**Comment by raveit65** [comment]
Beside from tab/space issue it looks ready to go for me.

**Comment by raveit65** [comment]
yeah, go ahead.

**Comment by lukefromdc** [comment]
OK, I will get on that tomorrow or the next day

**Comment by monsta** [comment]
First and second commits can be squashed...

The code in `panel-session-manager.c` seems to work, as I still can invoke logout and shutdown dialogs from the menu. Not sure how to test the changes in `panel-dbus-service.c`, in particular, the removal of reaction to `"NameOwnerChanged"`.

File: mate-panel/libpanel-util/panel-dbus-service.c [hunk 1]
```diff
@@ -54,6 +51,12 @@ static void panel_dbus_service_get_property (GObject    *object,
 					     guint       prop_id,
 					     GValue     *value,
 					     GParamSpec *pspec);
+void
+panel_dbus_service_on_signal (GDBusProxy *proxy,
```
Comment (monsta) [inline mate-panel/libpanel-util/panel-dbus-service.c]: This is a declaration of a function which doesn't exist. It does nothing as is. 

@yetist: did you plan to implement that function as a replacement for the removed `panel_dbus_service_on_name_owner_changed`?

**Comment by muktupavels** [comment]
Seems that `PanelDBusService` is used only by `PanelSessionManager`... You won't get `NameOwnerChanged` for it unless you know how to restart gnome-session without terminating your session. Also `GDBusProxy` handles that signal for you - https://gitlab.gnome.org/GNOME/glib/blob/master/gio/gdbusproxy.c#L1772.

**Comment by raveit65** [comment]
@yetist 
Can we finish the PR?
I want to avoid that your PR doesn't clean apply to master after some huge other open PRs, eg. wayland port.

**Comment by lukefromdc** [comment]
We really need to get this done soon, with all the other PRs coming through anything like this that is being held gets farther behind.

**Comment by yetist** [comment]
@raveit65 
Now add "NameOwnerChanged" signal process, and some indent, please review again, thank you.

**Comment by muktupavels** [comment]
Why "NameOwnerChanged" is needed? In case you get that signal won't you subscribe again to that signal?

**Comment by yetist** [comment]
> Why "NameOwnerChanged" is needed? In case you get that signal won't you subscribe again to that signal?

I'm not sure if there's a possibility of using mate-panel with other sessions, or switching / restarting / repalacing that session during use, and if that's the case, it's necessary to monitor the NemeOwnerChanged signal.

Mate-panel monitors the NameOwnerChanged signal on the session bus of org.freedesktop.DBUS, so when the change occurs, the connection does not change, so we don't need to subscribe the signal again.

**Comment by muktupavels** [comment]
On Thu, Nov 1, 2018 at 2:55 AM Wu Xiaotian <notifications@github.com> wrote:

> Why "NameOwnerChanged" is needed? In case you get that signal won't you
> subscribe again to that signal?
>
> I'm not sure if there's a possibility of using mate-panel with other
> sessions, or switching / restarting / repalacing that session during use,
> and if that's the case, it's necessary to monitor the NemeOwnerChanged
> signal.
>
Did you miss that I told you that `GDBusProxy` proxy handles that signal
for you?

> Mate-panel monitors the NameOwnerChanged signal on the session bus of
> org.freedesktop.DBUS, so when the change occurs, the connection does not
> change, so we don't need to subscribe the signal again.
>
Don't need, but that will happen, no? When you get that signal you will
call `panel_dbus_service_ensure_connection`, right? It will re-create proxy
and will subscribe, again, no? Or am I reading code wrong?

**Comment by yetist** [comment]
> On Thu, Nov 1, 2018 at 2:55 AM Wu Xiaotian ***@***.***> wrote: Why "NameOwnerChanged" is needed? In case you get that signal won't you subscribe again to that signal? I'm not sure if there's a possibility of using mate-panel with other sessions, or switching / restarting / repalacing that session during use, and if that's the case, it's necessary to monitor the NemeOwnerChanged signal.
> Did you miss that I told you that `GDBusProxy` proxy handles that signal for you?
> Mate-panel monitors the NameOwnerChanged signal on the session bus of org.freedesktop.DBUS, so when the change occurs, the connection does not change, so we don't need to subscribe the signal again.
> Don't need, but that will happen, no? When you get that signal you will call `panel_dbus_service_ensure_connection`, right? It will re-create proxy and will subscribe, again, no? Or am I reading code wrong?

Yes, subscribe signal code inside re-create proxy block is wrong, so I need move that code to other place or just drop that code about "NameOwenerchanged" signal?

**Comment by muktupavels** [comment]
From what I see I would say that handling `NameOwnerChanged` is not needed because `GDBusProxy` does that already for you, at least it looks like it does. Maybe best thing is to write small test case to verify that...

**Comment by raveit65** [comment]
I think there are more places where intends are not correct or a new code line doesn#t start under a bracket of the line before.
https://patch-diff.githubusercontent.com/raw/mate-desktop/mate-panel/pull/864.diff
But i am really tired to say people who knows much more of coding like me, they should work more cleaner.

**Comment by raveit65** [comment]
Btw. can you squash some commits and rebase the PR.
No need that intends fixes are a new commit.
You can do that with `git rebase -i <last-commit-id-before-PR>`

**Comment by muktupavels** [comment]
Well it looks like `dbus_connection` is also not needed!? You can use `g_dbus_proxy_new_for_bus_sync`, like you did in set-timezone.c.

**Comment by yetist** [comment]
> Well it looks like `dbus_connection` is also not needed!? You can use `g_dbus_proxy_new_for_bus_sync`, like you did in set-timezone.c.

Yes, I think so, just get proxy and then call dbus method in panel-session-manager.c. If do that, panel-dbus-service.{h,c} should removed.

**Comment by yetist** [comment]
@muktupavels @raveit65 
I droped panel-dbus-service.{h,c} files, get proxy and call dbus method in panel-session-manager.c now, is it OK?

**Comment by muktupavels** [comment]
Speaking about `PanelSessionManager`... Put `GDBusProxy` in `PanelSessionManager` struct! Don't use static here, not needed.

Also it looks like proxy could be `NULL`, but you use it in `g_dbus_proxy_call_sync` assuming it will never be `NULL`.

**Comment by yetist** [comment]
What about to write code in panel-action-button.c, and remove panel-session-manager.c file?

**Comment by muktupavels** [comment]
Is it only place where it is used? Anyway check that with MATE devs... I would probably keep it.

**Comment by yetist** [comment]
> Speaking about `PanelSessionManager`... Put `GDBusProxy` in `PanelSessionManager` struct! Don't use static here, not needed.
> 
> Also it looks like proxy could be `NULL`, but you use it in `g_dbus_proxy_call_sync` assuming it will never be `NULL`.

All done, waiting for review.

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 2]
```diff
@@ -42,136 +49,102 @@ panel_session_manager_class_init (PanelSessionManagerClass *klass)
...
+				       NULL,
+				       &error);
+	if (_ret == NULL) {
 		g_warning ("Could not ask session manager if shut down is available: %s",
 			   error->message);
 		g_error_free (error);
-
 		return FALSE;
+	} else {
+		g_variant_get(_ret, "(b)", &is_shutdown_available);
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: missing space before `(`...

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 3]
```diff
@@ -42,136 +49,102 @@ panel_session_manager_class_init (PanelSessionManagerClass *klass)
...
 }
 
 void
 panel_session_manager_request_logout (PanelSessionManager           *manager,
 				      PanelSessionManagerLogoutType  mode)
 {
-	GError *error;
-	DBusGProxy *proxy;
+	GError *error = NULL;
+	GVariant *_ret;
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: Why underscore is needed?

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 4]
```diff
@@ -42,136 +49,102 @@ panel_session_manager_class_init (PanelSessionManagerClass *klass)
...
-		return;
-	}
-
-	proxy = panel_dbus_service_get_proxy (PANEL_DBUS_SERVICE (manager));
-
-	if (!dbus_g_proxy_call (proxy, "Logout", &error,
-				G_TYPE_UINT, mode, G_TYPE_INVALID,
-				G_TYPE_INVALID) &&
-	    error != NULL) {
+	g_return_if_fail (manager->proxy != NULL);
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: `g_return_if_fail` is meant for programmer errors:
https://developer.gnome.org/glib/stable/glib-Warnings-and-Assertions.html#g-return-if-fail

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 5]
```diff
@@ -42,136 +49,102 @@ panel_session_manager_class_init (PanelSessionManagerClass *klass)
...
-					   "org.gnome.SessionManager");
-
-	error = NULL;
-	if (!panel_dbus_service_ensure_connection (PANEL_DBUS_SERVICE (obj),
-						   &error)) {
-		g_message ("Could not connect to session manager: %s",
-			   error->message);
+	GError *error = NULL;
+
+	manager->proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: Would not be better to move this to `ensure_proxy` or something like that? And use it in all functions that use proxy? Not sure if that can happen, but if proxy creating fails then you will never try again to create proxy. Old code would try to create proxy if it is `NULL`...

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 6]
```diff
@@ -22,18 +22,25 @@
...
-						   GObjectConstructParam *construct_properties);
+struct _PanelSessionManager {
+	GObject     parent;
+	GDBusProxy *proxy;
+};
+
+G_DEFINE_TYPE (PanelSessionManager, panel_session_manager, G_TYPE_OBJECT);
+
+static void
+panel_session_manager_finalize (GObject *object)
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: You need to chain-up in finalize... `G_OBJECT_CLASS (...)->finalize (...);`

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 7]
```diff
@@ -22,18 +22,25 @@
...
 
-static GObject *panel_session_manager_constructor (GType                  type,
-						   guint                  n_construct_properties,
-						   GObjectConstructParam *construct_properties);
+struct _PanelSessionManager {
+	GObject     parent;
+	GDBusProxy *proxy;
+};
+
+G_DEFINE_TYPE (PanelSessionManager, panel_session_manager, G_TYPE_OBJECT);
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: Trailing semicolon is not needed.

File: applets/clock/set-timezone.c [hunk 8]
```diff
@@ -171,75 +178,63 @@ free_data (gpointer d)
...
+			data->callback (data->data, NULL);
 	}
 }
 
 static void
 set_time_async (SetTimeCallbackData *data)
 {
-        DBusGConnection *bus;
-        DBusGProxy      *proxy;
+        GDBusProxy      *proxy;
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Unneeded space.

File: applets/clock/set-timezone.c [hunk 9]
```diff
@@ -171,75 +178,63 @@ free_data (gpointer d)
...
-						      set_time_notify,
-						      data, free_data,
-						      INT_MAX,
-						      /* parameters: */
-						      G_TYPE_INT64, data->time,
-						      G_TYPE_INVALID,
-						      /* return values: */
-						      G_TYPE_INVALID);
+		g_dbus_proxy_call (proxy,
+				"SetTime",
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Does not align correctly...

File: applets/clock/set-timezone.c [hunk 10]
```diff
@@ -171,75 +178,63 @@ free_data (gpointer d)
...
 				data->callback (data->data, error);
 			else
 				g_error_free (error);
-		}		
+		} else {
+			if (data->callback)
+				data->callback (data->data, NULL);
+		}
+	} else {
+		g_variant_get (variant, "()");
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: What this is supposed to do?

File: applets/clock/set-timezone.c [hunk 11]
```diff
@@ -171,75 +178,63 @@ free_data (gpointer d)
 }
 
 static void
-set_time_notify (DBusGProxy     *proxy,
-		 DBusGProxyCall *call,
-		 void           *user_data)
+set_time_notify (GObject *source_object,
+		GAsyncResult *res,
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Does not align correctly.

File: applets/clock/set-timezone.c [hunk 12]
```diff
@@ -171,75 +178,63 @@ free_data (gpointer d)
...
-			/* these errors happen because dbus doesn't
-			 * use monotonic clocks
-			 */	
-			g_warning ("ignoring no-reply error when setting time");
-			g_error_free (error);
-			if (data->callback)
-				data->callback (data->data, NULL);
-		}
-		else {
+	GError     *error = NULL;
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Align all variables or don't... Do not mix.

File: applets/clock/set-timezone.c [hunk 13]
```diff
@@ -171,75 +178,63 @@ free_data (gpointer d)
...
+	GDBusProxy *proxy;
+	GVariant   *variant;
+
+	proxy = get_bus_proxy ();
+	variant = g_dbus_proxy_call_finish (proxy, res, &error);
+	if (variant == NULL) {
+		if (error != NULL) {
 			if (data->callback)
 				data->callback (data->data, error);
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Are callback supposed to free error? Or we have memory leak here?

File: applets/clock/set-timezone.c [hunk 14]
```diff
@@ -28,74 +28,81 @@
...
-					   "/",
-					   "org.mate.SettingsDaemon.DateTimeMechanism");
-
-	dbus_g_proxy_begin_call_with_timeout (proxy,
-					      action,
-					      notify_can_do,
-					      callback, NULL,
-					      INT_MAX,
-					      G_TYPE_INVALID);
+	GDBusProxy      *proxy;
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Unneeded space.

File: applets/clock/set-timezone.c [hunk 15]
```diff
@@ -28,74 +28,81 @@
...
 #define CACHE_VALIDITY_SEC 2
 
 typedef  void (*CanDoFunc) (gint value);
 
 static void
-notify_can_do (DBusGProxy     *proxy,
-	       DBusGProxyCall *call,
-	       void           *user_data)
+notify_can_do (GObject *source_object,
+		GAsyncResult *res,
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Does not align correctly.

File: applets/clock/set-timezone.c [hunk 16]
```diff
@@ -28,74 +28,81 @@
...
-					      INT_MAX,
-					      G_TYPE_INVALID);
+	GDBusProxy      *proxy;
+
+	proxy = get_bus_proxy ();
+	if (proxy == NULL)
+		return;
+
+	g_dbus_proxy_call (proxy,
+			action,
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Not aligned.

File: applets/clock/set-timezone.c [hunk 17]
```diff
@@ -28,74 +28,81 @@
...
+		proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SYSTEM,
+				G_DBUS_PROXY_FLAGS_NONE,
+				NULL,
+				DATETIME_DBUS_NAME,
+				DATETIME_DBUS_PATH,
+				DATETIME_DBUS_NAME,
+				NULL,
+				&error);
+		if (proxy == NULL)
+		{
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: `{` should not be in new line.

File: applets/clock/set-timezone.c [hunk 18]
```diff
@@ -28,74 +28,81 @@
...
-	if (bus == NULL) {
-        	error = NULL;
-        	bus = dbus_g_bus_get (DBUS_BUS_SYSTEM, &error);
-        	if (bus == NULL) {
-                	g_warning ("Couldn't connect to system bus: %s", 
-				   error->message);
-                	g_error_free (error);
+	GError            *error = NULL;
+	static GDBusProxy *proxy = NULL;
+	if ( proxy == NULL) {
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: extra unneeded space before `proxy`.

File: applets/clock/set-timezone.c [hunk 19]
```diff
@@ -28,74 +28,81 @@
...
+		proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SYSTEM,
+				G_DBUS_PROXY_FLAGS_NONE,
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Alignment.

File: applets/clock/set-timezone.c (hunk 9 above)
Comment (muktupavels) [reply applets/clock/set-timezone.c]: My personal opinion is that tabs should not be used, only spaces. To keep existing style use combination of tab/tabs and spaces. For example:
```
[tab]my_function (param1,
[tab][spaaaaaaces]param2);
```

File: mate-panel/libpanel-util/panel-session-manager.c (hunk 6 above)
Comment (yetist) [reply mate-panel/libpanel-util/panel-session-manager.c]: ok, yes, I forgot it.

File: applets/clock/set-timezone.c (hunk 9 above)
Comment (yetist) [reply applets/clock/set-timezone.c]: Good. It looks like a better way.
Is there any possibility of using tools?

File: applets/clock/set-timezone.c [hunk 20]
```diff
@@ -28,74 +28,80 @@
...
 
 typedef  void (*CanDoFunc) (gint value);
 
 static void
-notify_can_do (DBusGProxy     *proxy,
-	       DBusGProxyCall *call,
-	       void           *user_data)
+notify_can_do (GObject *source_object,
+	       GAsyncResult *res,
+	       gpointer user_data)
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Still not aligned.

```
notify_can_do (GObject      *source_object,
               GAsyncResult *res,
               gpointer      user_data)
```

File: applets/clock/set-timezone.c [hunk 21]
```diff
@@ -171,75 +177,61 @@ free_data (gpointer d)
 }
 
 static void
-set_time_notify (DBusGProxy     *proxy,
-		 DBusGProxyCall *call,
-		 void           *user_data)
+set_time_notify (GObject *source_object,
+		 GAsyncResult *res,
+		 gpointer user_data)
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Same thing...

File: applets/clock/set-timezone.c (hunk 13 above)
Comment (muktupavels) [reply applets/clock/set-timezone.c]: Any comment on this?

File: applets/clock/set-timezone.c [hunk 22]
```diff
@@ -171,75 +177,61 @@ free_data (gpointer d)
...
+			data->callback (data->data, NULL);
 	}
 }
 
 static void
 set_time_async (SetTimeCallbackData *data)
 {
-        DBusGConnection *bus;
-        DBusGProxy      *proxy;
+        GDBusProxy *proxy;
```
Comment (muktupavels) [inline applets/clock/set-timezone.c]: Indentation...

File: mate-panel/libpanel-util/panel-session-manager.c [hunk 23]
```diff
@@ -42,136 +51,102 @@ panel_session_manager_class_init (PanelSessionManagerClass *klass)
...
+	manager->proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
+							G_DBUS_PROXY_FLAGS_NONE,
+							NULL,
+							"org.gnome.SessionManager",
+							"/org/gnome/SessionManager",
+							"org.gnome.SessionManager",
+							NULL,
+							&error);
+	if (manager->proxy == NULL)
+	{
```
Comment (muktupavels) [inline mate-panel/libpanel-util/panel-session-manager.c]: Should not be in new line.

File: mate-panel/libpanel-util/panel-session-manager.c (hunk 4 above)
Comment (muktupavels) [reply mate-panel/libpanel-util/panel-session-manager.c]: ?

**Comment by yetist** [comment]
@raveit65 
I'm sorry, I didn't handle it in time to keep you waiting.

Mixed space and tab can cause too many alignment problems. I use tab by default, but it looks like Mate is using space?

**Comment by raveit65** [comment]
You're using tab size=8, right?
I am using tab size=4 in pluma, which is default.
The result of this mismatch is that the resulting patch of your PR looks weird in my pluma.
Try it out for yourself and change tab size, maybe than you understand the problem with mixing spaces and tabs in one line.

**Comment by raveit65** [comment]
I squashed all in one commit, which is better for a meaningful git history.
merged
https://github.com/mate-desktop/mate-panel/commit/a34f531da4f03eecfc38414295ebe70742dfad71
But next time i won't merge PRs any more with tab vs. spaces problems.

**Comment by muktupavels** [comment]
@raveit65 did you saw more indentation problems? Last changes looks more or less good to me. There was two places where function parameters could be aligned. But there are one change that I don't like - he removed `g_return_val_if_fail`, but should have replaced it with early return probably with warning that proxy is not available or something like that.

I probably was not clear enough, my fault. Right now if proxy is NULL it will crash, it could crash also with `g_return_val_if_fail`:

> If G_DISABLE_CHECKS is defined then the check is not performed. You should therefore not depend on any side effects of expr .

**Comment by raveit65** [comment]
wtf.....
@muktupavels 
Indentations was OK if you use tab size = 8 in editor.

@yetist 
Can you take a look again to fix last concern?

**Comment by muktupavels** [comment]
I was look at diff here in github... Well if file is already using tabs you typically want to stick to same style. Looking again at diff I see he wrongly aligned arguments - for arguments you want use tabs (same time you used for first line) and everything else should be spaces.

**Comment by raveit65** [comment]
Resulting commit looks OK for me if open in browser
https://github.com/mate-desktop/mate-panel/commit/a34f531da4f03eecfc38414295ebe70742dfad71
Or if you use tab size =8 in editor.
Any way, a bigger problem is the issue which you mentioned.
How can i fix that?

**Comment by muktupavels** [comment]
In simplest form:
`if (!manager->proxy) return;`

Or if you want to print debug or warning message:
```
if (!manager->proxy) {
    g_debug / g_warning ("...");
    return;
}
```

**Comment by lukefromdc** [comment]
Just how bad are the tab/indent issues here? If we are holding up this whole PR over it, we are also holding up the Wayland port and risking further diff growth as other things get worked on. I propose we get this merged and open a new PR to work on the indent issues

**Comment by lukefromdc** [comment]
OK, I see this is merged and the remaining indent issues can be dealtt with separately. That needs to be fixed, but the way is now open for the wayland work

**Comment by yetist** [comment]
> In simplest form:
> `if (!manager->proxy) return;`
> 
> Or if you want to print debug or warning message:
> 
> ```
> if (!manager->proxy) {
>     g_debug / g_warning ("...");
>     return;
> }
> ```

I tested on my system with follow code:
```
#include <gio/gio.h>
int main(int argc, char** argv)
{
    GDBusProxy *proxy;
    GError *error = NULL;
    proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
            G_DBUS_PROXY_FLAGS_NONE,
            NULL,
            "org.gnome.SessionManager1",    // Here use wrong name, wrong path
            "/org/gnome/SessionManager2",   // and wrong interface for test,
            "org.gnome.SessionManager3",    // like mate-session not running.
            NULL,
            &error);
    if (proxy == NULL) {
        g_warning ("Unable to connect session manager daemon: %s\n", error->message);
        g_error_free (error);
    } else {
        g_print ("proxy is not null\n");
    }
    return 0;
}
```
But I get output `proxy is not null` always,  so I think no need to check `manager->proxy == NULL` when call dbus method,  maybe I was wrong.

**Comment by yetist** [comment]
More test:
```
$ cat test.c 
#include <gio/gio.h>
int main(int argc, char** argv)
{
    GDBusProxy *proxy;
    GVariant *ret;
    GError *error = NULL;
    gboolean is_shutdown_available;

    proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
            G_DBUS_PROXY_FLAGS_NONE,
            NULL,
            "org.gnome.SessionManager1",    // Here use wrong name, wrong path
            "/org/gnome/SessionManager2",   // and wrong interface, to test
            "org.gnome.SessionManager3",    // mate-session not running.
            NULL,
            &error);
    if (proxy == NULL) {
        g_warning ("Unable to connect session manager daemon: %s\n", error->message);
        g_error_free (error);
    } else {
        g_print ("proxy is not null\n");
    }

    g_assert(error == NULL);

    ret = g_dbus_proxy_call_sync (proxy, "CanShutdown",
            g_variant_new ("()"),
            G_DBUS_CALL_FLAGS_NONE,
            -1,
            NULL,
            &error);
    if (ret == NULL) {
        g_print ("ERROR: %s\n", error->message);
        g_error_free (error);
    } else {
        g_variant_get (ret, "(b)", &is_shutdown_available);
        g_variant_unref (ret);
    }

    g_print("shutdown available is %d\n", is_shutdown_available);

    return 0;
}
$ gcc -o test test.c `pkg-config --cflags --libs gio-2.0`
$ ./test 
proxy is not null
ERROR: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.gnome.SessionManager1 was not provided by any .service files
shutdown available is 0
```
If `proxy` is not null,  atfer call `g_dbus_proxy_call_sync` , will print warning message, if proxy is NULL,  call `g_dbus_proxy_call_sync` will crash.
But the question is that after call `g_dbus_proxy_new_for_bus_sync`, it always return no null, is that ok?  At lease on my system it looks like this, maybe my system has error, or I was wrong.
Does anyone have time to test the above code?

**Comment by muktupavels** [comment]
`g_dbus_proxy_new_for_bus_sync` is documented that it can return `NULL`... So I would say that you should handle that! Anyway, best thing is to read source code not guess when it can be `NULL`.

**Comment by raveit65** [comment]
i got same results with test.c
```
[rave@mother Schreibtisch]$ ./test 
proxy is not null
ERROR: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.gnome.SessionManager1 was not provided by any .service files
shutdown available is 0
```

**Comment by raveit65** [comment]
Same result in multiuser-target without running X (runlevel 3) after a `killall dbus-daemon`
```
[rave@mother Schreibtisch]$ ./test 
proxy is not null
ERROR: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.gnome.SessionManager1 was not provided by any .service files
shutdown available is 0
```

**Pull request closed without merging**